module/state_synchronization/requester/checkstore.go

```diff
@@ -0,0 +1,193 @@
+	// heights. All data should be present, otherwise data was deleted or corrupted on disk.
+	for height := c.startBlockHeight; height <= lastHeight; height++ {
+		if ctx.Err() != nil {
+			return nil
+		}
+
+		header, err := c.headers.ByHeight(height)
+		if err != nil {
+			return fmt.Errorf("failed to get block for height %d: %w", height, err)
+		}
+
+		result, err := c.results.ByBlockID(header.ID())
+		if err != nil {
+			return fmt.Errorf("failed to lookup execution result for block %d: %w", header.ID(), err)
+		}
+
+		checkLogger := c.log.With().
+			Uint64("height", height).
+			Str("block_id", header.ID().String()).
+			Str("execution_data_id", result.ExecutionDataID.String()).
+			Logger()
+
+		exists, err := c.checkExecutionData(ctx, result.ExecutionDataID, checkLogger)
+
+		if err != nil {
+			return err
+		}
+
+		if !exists {
+			checkLogger.Debug().Msg("redownloading missing blob")
+			// blocking download of any missing block that should exist in the datastore
+			// TODO: ideally, this should push a job into the blockConsumer's queue for the worker
+			// pool to consume. However, the jobqueue doesn't currently have a convenient way to
+			// push work into the queue.
+			err = c.downloadExecutionData(ctx, header.ID(), height)
+			if err != nil && err != ctx.Err() {
+				return fmt.Errorf("failed to redownload execution data for height %d: %w", height, err)
+			}
+		}
+	}
+
+	c.log.Info().
+		Uint64("start_height", c.startBlockHeight).
+		Uint64("end_height", lastHeight).
+		Uint64("total_checked", lastHeight-c.startBlockHeight+1).
+		Msg("datastore check successful")
+
+	return nil
+}
+
+// checkExecutionData checks the datastore for a specific ExecutionData blob
+func (c *datastoreCheckerImpl) checkExecutionData(ctx irrecoverable.SignalerContext, rootID flow.Identifier, logger zerolog.Logger) (bool, error) {
+	invalidCIDs, ok := c.eds.Check(ctx, rootID)
+
+	// Check returns a list of CIDs with the corresponding errors encountered while retrieving their
+	// data from the local datastore.
+
+	if ok {
+		logger.Debug().Msg("check ok")
+		return true, nil
+	}
+
+	// One or more of the blobs are invalid. Inspect the errors to determine if any are corrupt or
+	// have unexpected errors. Delete any corrupt blobs, and record unexpected errors. When the
+	// ExecutionData is requested again via the ExecutionDataService, any blobs that are missing
+	// from the blobstore will be redownloaded, and any that exist are ignored.
+
+	var errs *multierror.Error
+	for _, invalidCID := range invalidCIDs {
+		cid := invalidCID.Cid
+		err := invalidCID.Err
+
+		// Not Found. Missing blobs will be redownloaded
+		if errors.Is(err, blockservice.ErrNotFound) {
+			logger.Info().Str("cid", cid.String()).Msg("blob not found")
+			continue
+		}
+
+		// The blob's hash didn't match. This is a special case where the data was corrupted on
+		// disk. Delete and redownload.
+		if errors.Is(err, blockstore.ErrHashMismatch) {
```

Comment:
I don't think data corruption case is due to storage's fault, instead should be malicious's upload. 

I think Bitswap will probably whether check hash matches before storing them to storage (disk). And the storage layer (badger) would not return corrupted data to upper layer, if Bitswap only only store data when hash matches.

When HashMismatch happens, there is no point to continue keep this data in the database. I think data service should just delete this data as soon as found HashMismatch.

That means the data service's Get method should delete the downloaded but HashMismatch data, and return an error to the caller. If the caller call `Get`, it will just download again. 

ExecutionDataService could internally keep a counter of how many times it has retried to download the same rootCID, and decide whether to download again. 

I think this

Reply:
I also don't quite understand how `HashMismatch` would return from storage given we never store mismatched data to storage.

We will get HashMismatch data after downloaded it, and before storing it. But when reading the storage, we should not get `HashMismatch` error, right? Do I miss anything?

Reply:
Yea, so if a blobs hash didn't match when it was originally received, Bitswap would not store it, and the requester  would eventually retry.

It's possible the data could later become corrupted on disk (e.g. a bitflip or some other accidental corruption). In the event of data corruption, we want to be able to redownload the corrupted data without redownloading all of the data. That's what the datastore checker does. It does it by using an ExecutionDataService that's configure to not use Bitswap, so it does not attempt to download missing data from the network. It also uses a blobstore which is configured to hash all of the blobs on read, to verify that their CID still match their hash. If not, it returns `blockstore.ErrHashMismatch` from `GetBlob` (note the singular naming). It's effectively checkdisk for the ExecutionData blobstore.

The ExecutionDataService's `Get` method uses the blobstore's `GetBlobs` (note the plural naming) to get a group of blobs from storage. The way that method works is it either returns the data via a channel, or times out. So it can't return `blockstore.ErrHashMismatch` for a specific blob.

Reply:
Is the bitflip case from the Bitswap or from storage? 

If the accidental corruption could happen anytime, if it happens after the check, then wouldn't we still read corrupted data?

Is there any reference about the cause of the bitflip case?

Reply:
bitflip would be from storage (possibly a bad disk). Yea, it if happened after the check, we'd read bad data. I think there's an outstanding action item for us to understand how bitswap would react to reading bad data when responding to requests. We may want to also enable the `HashOnRead` option on the bitswap ExecutionDataService, but need to do some testing first to check on performance.

Reply:
If that's something we have to worry about, isn't the existing protocol state having the same risk? They are both built on top of badger db.

If this risk is real, it's better understand it better first.

Reply:
that's fair. I'll remove it and we can add it back later if it becomes a problem.

Reply:
update in 7983803